### PR TITLE
chore: isolate custom example app to be injected on matter SDK

### PR DESCRIPTION
**Pull request by locnnil**
* add: Missing GN build system files
* add: Missing source and header files
* clean: Remove non-needed pigweed configs on GN files.
* refact: update app name to avoid conflicts on SDK.

### CODE REVIEW

File: snap/snapcraft.yaml [hunk 1]
```diff
@@ -115,36 +115,36 @@ parts:
...
+  #   override-build: |
+
+  #     if [[ $SNAP_ARCH == "arm64" ]]; then
+  #       # Download and unzip the prebuilt ZAP (Zigbee Cluster Library configuration tool and generator) binary for the ARM64 architecture
+  #       wget --no-verbose https://github.com/project-chip/zap/releases/download/$ZAP_VERSION/zap-linux-$SNAP_ARCH.zip
+  #       unzip -o zap-linux-$SNAP_ARCH.zip
+
+  #       # Define the environment needed for the app build
+  #       echo "export ZAP_INSTALL_PATH=$PWD" >> env
+  #     fi
```
Comment (farshidtz) [inline snap/snapcraft.yaml]: This change is about #61 and unrelated to the chore task. Please split into another PR.

File: snap/snapcraft.yaml [hunk 2]
```diff
@@ -115,36 +115,36 @@ parts:
...
 
   lighting:
-    after: [libgpiod, test-blink, connectedhomeip, zap]
+    after: [libgpiod, test-blink, connectedhomeip]
     plugin: nil
-    source: app
-    build-environment:
-      - ZAP_ENV: ../../zap/build/env
+    source: .
```
Comment (farshidtz) [inline snap/snapcraft.yaml]: Why is the `source` path changed? Setting the source to the whole project directory will result in invalidated part's cache every time unrelated code, such as tests change.

File: snap/snapcraft.yaml [hunk 3]
```diff
@@ -173,18 +173,19 @@ parts:
...
+      # See https://github.com/project-chip/connectedhomeip/issues/32925
+      cp -vr $CRAFT_PART_BUILD/app examples/
 
-      cd examples/lighting-app/linux
+      cd examples/app/linux
       gn gen out/build
       ninja -C out/build
 
-      ldd out/build/chip-lighting-app
+      ls out/build
```
Comment (farshidtz) [inline snap/snapcraft.yaml]:
```suggestion
      # List shared objects
```

File: snap/snapcraft.yaml [hunk 4]
```diff
@@ -173,18 +173,19 @@ parts:
       source scripts/activate.sh
       set -u
 
-      # Copy and replace the application files
-      cp -vr $CRAFT_PART_BUILD/lighting-common/* examples/lighting-app/lighting-common/
-      cp -vr $CRAFT_PART_BUILD/linux/* examples/lighting-app/linux
+      # Add the App to be build from CHIP SDK's build system
+      # See https://github.com/project-chip/connectedhomeip/issues/32925
+      cp -vr $CRAFT_PART_BUILD/app examples/
```
Comment (farshidtz) [inline snap/snapcraft.yaml]: I suggest giving the `app` directory a more distinct name, such as `gpio-app` to make it clearer that app/linux/BUILD.gn (line 46) is using the common code taken from this repo. This should also be consistent with the binary name.

```suggestion
      cp -vr $CRAFT_PART_BUILD/app examples/gpio-app
```

**Review by farshidtz — CHANGES_REQUESTED** [review]
Thanks for the changes.
Please base your code on a tagged version, such as [v1.3.0.0](https://github.com/project-chip/connectedhomeip/releases/tag/v1.3.0.0) to ensure better stability. It is currently based on https://github.com/project-chip/connectedhomeip/commit/8a4dffc5d

File: snap/snapcraft.yaml (hunk 2 above)
Comment (locnnil) [reply snap/snapcraft.yaml]: Because I'm using the whole `app` directory to be copied to inside the `examples` directory on chip SDK.

File: app/README.md [hunk 5]
```diff
@@ -1,5 +1,5 @@
 # App
 
-The files under this directory get copied into CHIP project's [examples/lighting-app](https://github.com/project-chip/connectedhomeip/tree/master/examples/lighting-app) 
-to inject GPIO control functionality and removes the non-functional occupancy sensor from the upstream example. 
-The reason for this approach is to maintain the simplicity of this example and rely on the existing upstream build configurations.
+This directory is copied into CHIP project's 
+[examples/lighting-app](https://github.com/project-chip/connectedhomeip/tree/master/examples/lighting-app),
+the reason for that is described on this Chip projects [upstream issue](https://github.com/project-chip/connectedhomeip/issues/32925).
```
Comment (farshidtz) [inline app/README.md]:
```suggestion
To build, this directory should be copied into CHIP project's 
[examples](https://github.com/project-chip/connectedhomeip/tree/master/examples) directory.
The reason for that is described [in this issue](https://github.com/project-chip/connectedhomeip/issues/32925).
```

File: snap/snapcraft.yaml [hunk 6]
```diff
@@ -134,7 +134,7 @@ parts:
       fi
 
   lighting:
-    after: [libgpiod, test-blink, connectedhomeip, zap]
+    after: [libgpiod, test-blink, connectedhomeip]
```
Comment (farshidtz) [inline snap/snapcraft.yaml]:
```suggestion
    after: [libgpiod, test-blink, connectedhomeip, zap]
```
This should be reverted, since zap is still a dependency for arm64's build. It should be set even if it works without it to have a predictable execution order of these parts.

**Review by farshidtz — CHANGES_REQUESTED** [review]

**Review by farshidtz — APPROVED** [review]
Looks good, thanks.
